google/cloud/sql/connector/rate_limiter.py

```diff
@@ -0,0 +1,114 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import asyncio
+import math
+import time
+from typing import Optional, Type
+from types import TracebackType
+
+
+class AsyncRateLimiter(object):
+    """
+    An asyncio-compatible rate limiter which uses the Token Bucket algorithm
+    (https://en.wikipedia.org/wiki/Token_bucket) to limit the number of function calls over a time interval using an event queue.
+
+    :type burst_size: int
+    :param: burst_size:
+        The max number of function calls that can occur at the same time
+        Default: 1
+
+    :type interval: int
+    :param: interval:
+        The period of time over which a number of calls equal to burst_size are allowed to complete. Default: 60s
+
+    :type queue_size: int
+    :param: interval:
+        The number of tasks that are allowed to be queued if rate limiter cannot be acquired. If exceeded, a QueueFull error will be thrown. A queue size of zero means the queue length has no upper bound. If setting queue size to a value greater than zero, queue_size must be greater than or equal to burst_size
+        Default: 0
+
+    :type loop: asyncio.AbstractEventLoop
+    :param: loop:
+        The event loop to use. If not provided, the default event loop will be used.
+
+
+    """
+
+    def __init__(
+        self,
+        burst_size: int = 1,
+        interval: float = 60,
+        loop: asyncio.AbstractEventLoop = None,
+    ) -> None:
+        self.interval = interval
+        self.burst_size = burst_size
+        self._loop = loop or asyncio.get_event_loop()
+        self._tokens = asyncio.BoundedSemaphore(burst_size, loop=self._loop)
+        self._last_token_update = time.time()
+
+    def _add_tokens(self) -> None:
+        """
+        Calculates how much time has passed since the last leak and removes the
+        appropriate amount of events from the queue.
+        Leaking is done lazily, meaning that if there is a large time gap between
+        leaks, the next set of calls might be a burst if burst_size > 1
+        """
+        time_elapsed = time.time() - self._last_token_update
+        amount = math.floor(time_elapsed * self.burst_size / self.interval)
+        # don't allow more than burst_size to leak at once
+        leak_amount = min(amount, self.burst_size)
+        if amount > 0:
+            self._last_token_update = time.time()
+        for _ in range(leak_amount):
+            try:
+                self._tokens.release()
+            except ValueError:
+                # we want to keep adding tokens until the semaphore reaches its upper bound
+                break
+
+    def _token_acquired(self, waiter_task: asyncio.Task) -> bool:
+        """
+        Releases tasks out of the queue if enough time has elapsed.
+        """
+        self._add_tokens()
+        return waiter_task.done()
```

Comment:
I'm not following what this is doing.

Reply:
the waiter task is waiting to acquire the semaphore (i.e. acquire a token) . After adding tokens, this checks to see if the task has gotten a token, and if it has, it breaks out of the loop in `acquire`. I'm using the loop so that we can periodically check if enough time has passed to add another token.

Reply:
I think maybe we should chat about this, because I think we should be able to eliminate some of the complexity here. Specifically it feels like the logic should be something like: 

```
async with lock
    calculate updated token count
    if tokens <= 0:
        wait until next token is ready
        calculate updated token count
    wait until time the next token is ready
    update token count
    tokens -= 1
        // success!
```